repair/row_level.cc

```diff
@@ -676,6 +676,28 @@ void flush_rows(schema_ptr s, std::list<repair_row>& rows, lw_shared_ptr<repair_
     }
 }
 
+struct global_reader_permit_meta {
+    std::vector<foreign_ptr<lw_shared_ptr<semaphore_units<>>>> permits{smp::count};
+};
+
+future<global_reader_permit_meta> get_global_reader_permit(repair_service& rs, unsigned shard, bool all_shards) {
+    global_reader_permit_meta meta;
+    co_await rs.container().invoke_on(smp::count -1, [&meta, shard, all_shards] (repair_service& rs) -> future<> {
+        co_await with_semaphore(rs.lock_sem(), 1, [&rs, &meta, shard, all_shards] () -> future<> {
+            co_await rs.container().invoke_on_all([&meta, shard, all_shards] (repair_service& rs) -> future<> {
+                if (all_shards || shard == this_shard_id()) {
```

Comment:
Isn't this_shard_id() always smp::count - 1 because of the outer invoke_on call?
I'm afraid I don't understand what this function is trying to do to understand if it's doing it correctly or not. A comment could have helped.

Reply:
if all_shards == true, it means the caller wants to take the lock on all shards.

if all_sahrds == false, then a single shard is going to take the shard.

The if check above does exactly this logic.

Reply:
But if the concept of "global permit" means "a permit on all shards", then what does a "global permit with all_shards=false" mean?
Maybe it means I misunderstood what a "global permit" means?

Reply:
I still don't understand. What is this_shard_id() here, am I wrong that it's always smp::count-1?
If so, it means we always take the reader semaphore from smp::count-1 instead of the calling shard. Why? Don't we need to take it on the same shard that will actually be doing the reading?

Reply:
The check is under "invoke_on_all"  ;-)

Reply:
You're right, I missed that. So now I'm confused, why we need to do invoke_on_all in that case ;-)